[CSS Fonts] Implement the from-font value for font-size-adjust

shivamidow opened the pull request:
#### 0f2621d762162b939c8959ba3aa4f4a3af8452df
<pre>
[CSS Fonts] Implement the from-font value for font-size-adjust
<a href="https://bugs.webkit.org/show_bug.cgi?id=254790">https://bugs.webkit.org/show_bug.cgi?id=254790</a>

Reviewed by Myles C. Maxfield and Tim Nguyen.

The &apos;from-font&apos; value is newly added to CSS Font Module Level 5 [1], which allows
the browser to automatically determine a font-size-adjust value based on the primary font.
This change implements it.

The following tests have merged into the latest WPT [2] and are backported in this patch.

Test:
    imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-013.html
    imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed.html
    imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-invalid.html
    imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid.html

[1] <a href="https://www.w3.org/TR/css-fonts-5/#valdef-font-size-adjust-from-font">https://www.w3.org/TR/css-fonts-5/#valdef-font-size-adjust-from-font</a>
[2] <a href="https://github.com/web-platform-tests/wpt/pull/39380">https://github.com/web-platform-tests/wpt/pull/39380</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-013-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-013.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontSizeAdjust):
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
(WebCore::FontSizeAdjust::operator== const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontSizeAdjust):
* Source/WebCore/style/StyleFontSizeFunctions.cpp:
(WebCore::Style::aspectValueOfPrimaryFont):
* Source/WebCore/style/StyleFontSizeFunctions.h:

Canonical link: <a href="https://commits.webkit.org/262800@main">https://commits.webkit.org/262800@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02d387ca4bc4b51ee45b471e39c3a7c06890f466

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2656 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2659 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4064 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3037 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2621 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2799 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2754 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2339 "115 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2680 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2375 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3828 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/625 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2360 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2220 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2737 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2377 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/3581 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2411 "Passed tests") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2416 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2363 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/654 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2555 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid.html:
```diff
@@ -2,22 +2,30 @@
 test_valid_value('font-size-adjust', 'ic-width 0.9');
 test_valid_value('font-size-adjust', 'ic-height 0.9');
+
+test_valid_value('font-size-adjust', 'from-font');
+test_valid_value('font-size-adjust', 'ex-height from-font', 'from-font');  // default basis 'ex' omitted from serialization
+test_valid_value('font-size-adjust', 'cap-height from-font');
+test_valid_value('font-size-adjust', 'ch-width from-font');
+test_valid_value('font-size-adjust', 'ic-width from-font');
+test_valid_value('font-size-adjust', 'ic-height from-font');
```
would be nice to test some invalid combinations involving `from-font` as well.

shivamidow replied:
Would you say more? Do you mean a typo error, something like 'font-from'? Or else, extra value cases like `font-size-adjust: from-font 0.5` or `font-size-adjust: ex-height 0.5 from-font`?

nt1m replied:
Extra value cases, like the ones you suggested, or `font-size-adjust: from-font from-font` etc. These are usually common parsing mistakes

shivamidow replied:
O.K. I added some invalid tests to imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-invalid.html

nt1m replied:
Thanks! Please make sure to export them to WPT whenever you can, otherwise changes will be lost at the next import.

shivamidow replied:
Sure. Will do!

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1516,24 +1517,34 @@ inline FontVariationSettings BuilderConverter::convertFontVariationSettings(Buil
 
-        ASSERT(primitiveValue.isNumber());
-        return { FontSizeAdjust::Metric::ExHeight, primitiveValue.floatValue() };
+        FontSizeAdjust::Metric defaultMetric = FontSizeAdjust::Metric::ExHeight;
```
```suggestion
        auto defaultMetric = FontSizeAdjust::Metric::ExHeight;
```

shivamidow replied:
Fixed.

shivamidow committed:
[CSS Fonts] Implement the from-font value for font-size-adjust
https://bugs.webkit.org/show_bug.cgi?id=254790

Reviewed by Myles C. Maxfield and Tim Nguyen.

The 'from-font' value is newly added to CSS Font Module Level 5 [1], which allows
the browser to automatically determine a font-size-adjust value based on the primary font.
This change implements it.

The following tests have merged into the latest WPT [2] and are backported in this patch.

Test:
    imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-013.html
    imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed.html
    imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-invalid.html
    imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid.html

[1] https://www.w3.org/TR/css-fonts-5/#valdef-font-size-adjust-from-font
[2] https://github.com/web-platform-tests/wpt/pull/39380

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-013-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-013.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontSizeAdjust):
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
(WebCore::FontSizeAdjust::operator== const):
(WebCore::operator<<):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontSizeAdjust):
* Source/WebCore/style/StyleFontSizeFunctions.cpp:
(WebCore::Style::aspectValueOfPrimaryFont):
* Source/WebCore/style/StyleFontSizeFunctions.h:

Canonical link: https://commits.webkit.org/262800@main
